vehicle/mkv/software/throttle/throttle.c

```diff
@@ -1,738 +1,256 @@
-	gThrottle2In = map2;
-
-	gThrottle1Out = gThrottle1In;
-	gThrottle2Out = gThrottle2In;
-
-	/*
-	   //---------Map based on drive mode------------
-	   if (gDriveMode == 0) { //error, default to standard
-	   gThrottle1Out = gThrottle1In;
-	   gThrottle2Out = gThrottle2In;
-	   }
-	   else if (gDriveMode == 1) { //standard
-	   gThrottle1Out = gThrottle1In;
-	   gThrottle2Out = gThrottle2In;
-	   }
-	   else if (gDriveMode == 2) { //acceleration
-	   gThrottle1Out = gThrottle1In;
-	   gThrottle2Out = gThrottle2In;
-	   }
-	   else if (gDriveMode == 3) { //skid pad
-	   if (gThrottle1In < 128) {
-	    gThrottle1Out = gThrottle1In;
-	    gThrottle2Out = gThrottle2In;
-	   }
-	   else {
-	    gThrottle1Out = 128 + ((gThrottle1In - 128) >> 1);
-	    gThrottle2Out = 128 + ((gThrottle2In - 128) >> 1);
-	   }
-	   }
-	   else if (gDriveMode == 4) { //autocross
-	   gThrottle1Out = gThrottle1In;
-	   gThrottle2Out = gThrottle2In;
-	   }
-	   else if (gDriveMode == 5) { //endurance
-	   gThrottle1Out = gThrottle1In;
-	   gThrottle2Out = gThrottle2In;
-	   }
-	 */
-}
-
-void storeThrottle(void) {
-	readPots();
-	mapThrottle();
-	checkPlausibility();
 
-	uint16_t new = (gThrottle1Out + gThrottle2Out) >> 1;
-	uint8_t old = gThrottleArray[gThrottleArrayIndex];
-
-	gRollingSum -= old;
-	gRollingSum += new;
-
-	gThrottleArray[gThrottleArrayIndex] = new;
-
-	gThrottleArrayIndex = gThrottleArrayIndex + 1;
-
-	if (gThrottleArrayIndex == ROLLING_AVG_SIZE) {
-		gThrottleArrayIndex = 0;
-	}
-
-	// sprintf(uart_buf, "NEW: %u", new);
-	// LOG_println(uart_buf, strlen(uart_buf));
-	// sprintf(uart_buf, "OLD: %u", old);
-	// LOG_println(uart_buf, strlen(uart_buf));
+#include "libs/gpio/api.h"
+#include "throttle_config.h"
+#include "utils/timer.h"
+#include "vehicle/mkv/software/throttle/can_api.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#define ROLLING_AVG_SIZE          8
+#define IMPLAUSIBILITY_TIMEOUT_MS 100
+
+uint8_t deviation_counter = 0; // keeps track of elapsed time of >= 10%
+                               // deviation in mapped throttle values
+
+// volatile uint8_t gFlag = 0x00; // Global Flag. Tells us that gFlag will
+// change and to assign it to an unsigned 8bit integer volatile bool send_can =
+// false;
+typedef enum {
+    THROTTLE_IDLE = 0,
+    THROTTLE_RUN,
+    THROTTLE_IMPLAUSIBILITY,
+    THROTTLE_FAULT
+} throttle_state_e;
+
+// copied from air.c for now, but really should be made available via CAN api in
+// future
+typedef enum {
+    INIT = 0,
+    IDLE,
+    SHUTDOWN_CIRCUIT_CLOSED,
+    PRECHARGE,
+    TS_ACTIVE,
+    DISCHARGE,
+    FAULT,
+} air_state_e;
+
+typedef struct {
+    throttle_state_e state;
+    volatile bool send_can;
+    bool flag_implausibility;
+    uint16_t throttle0_mapped;
+    uint16_t throttle1_mapped;
+    uint16_t throttle_values[8];
+    uint16_t rolling_sum;
+    uint8_t implausibility_counter;
+    bool brake_pressed;
+    bool ready_to_drive;
+    air_state_e air_state;
+} throttle_state_s;
```

Comment:
I don't think a typedef is needed here. Maybe just something like

```c
struct throttle_state {
    ...
    ...
} throttle_state = { 0 };
```

This will also create the variable throttle state and initialize everything to zero.

Reply:
I feel luke-warm about having a single struct to contain all state, buuuut I don't think it's a huge deal? I think you should add comments for each member of the struct about what it is though, unless it's obvious. Also, group like-things together, like flag_implausibility and implausibility_counter.